### PR TITLE
Remove all env

### PR DESCRIPTION
**Pull request by DustinMoriarty**
# Pull Request Check List

Resolves: #3208

* Add support for removing all virtualenvs in the CLI.
* Add support for removing multiple virtualenvs in the CLI.
* Add tests for supported features.
* Revise CLI help to document supported features.

### CODE REVIEW

**Comment by max-sixty** [comment]
Awesome — thanks @DustinMoriarty !

File: poetry/console/commands/env/remove.py
Comment (abn) [inline poetry/console/commands/env/remove.py:30]: Might want to handle ambiguity condition where `all` and `python` is specified.

File: poetry.lock
Comment (abn) [inline poetry.lock:26]: Not sure why the lock file has been updated in this changeset. These changes should be dropped.

File: poetry/console/commands/env/remove.py
Comment (abn) [inline poetry/console/commands/env/remove.py:10]:
```suggestion
    description = "Removes specific virtual environments associated with the project."
```

While we are at it might as well improve the wording.

File: poetry/console/commands/env/remove.py
Comment (abn) [inline poetry/console/commands/env/remove.py:15]:
```suggestion
            "The python executable(s) to remove the virtual environment for.",
```

File: poetry/console/commands/env/remove.py
Comment (abn) [inline poetry/console/commands/env/remove.py:21]:
```suggestion
        option("all", description="Remove all managed virtual environments associated with the project."),
```

**Review by abn — CHANGES_REQUESTED** [review]
In general looking good, minor suggestions.

File: poetry.lock
Comment (DustinMoriarty) [reply poetry.lock:26]: Will do.

File: poetry/console/commands/env/remove.py
Comment (DustinMoriarty) [reply poetry/console/commands/env/remove.py]: Good point. I reordered the function so that it removes the python versions from the python argument and then if --all is provided then it removes any remaining environments. I think that will make the behavior more clear in the combined case. There is also a new test to validate this behavior.

**Comment by venaturum** [comment]
This looks good to go right?

**Comment by sinoroc** [comment]
> This looks good to go right?

Looks like it might need docs update as well.

**Review by steveYeah — APPROVED** [review]
Nice! This will also solve [this issue](https://github.com/python-poetry/poetry/issues/3617), so I'm keen to get this merged 😄  👍🏻  (Once the docs have been added, of course)

**Comment by DustinMoriarty** [comment]
@sinoroc : Which docs need to be updated?

**Comment by DustinMoriarty** [comment]
Ok. I see perhaps we could update [managing-environments.md](https://github.com/python-poetry/poetry/blob/master/docs/docs/managing-environments.md) to show the new functionality.  The old docs would not be  wrong since the change is back compatible. However, we can streamline the documentation for env remove to show multiple lines removed at once. Am I understanding your suggestion right @sinoroc ?

**Comment by sinoroc** [comment]
Yes. I had this in mind:
* https://python-poetry.org/docs/cli/#env
but it actually is just a link to:
* https://python-poetry.org/docs/managing-environments/
So in the end, as you said, I guess docs should go to:
* https://github.com/python-poetry/poetry/blob/master/docs/docs/managing-environments.md
At least that would be my recommendation (I am not a maintainer, so I do not have the last word on this).

**Comment by DustinMoriarty** [comment]
Got it. I'll see if I have some time to take a look this weekend.

**Comment by DustinMoriarty** [comment]
@sinoroc, @venaturum, @steveYeah  or @abn : I just went to update the docs as discussed as well as update the branch to the latest master. I am getting this failure with the FreeBSD build that does not look like it has anything to do with the PR. I think I had this problem before. It was odd because it was intermittent. Previously I would make some minor change, push and it would go away.

**Comment by DustinMoriarty** [comment]
Here is the error: 

```bash
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/cirrus-ci-build/.tox/py/lib/python3.8/site-packages/ptyprocess-0.7.0.dist-info'
```

File: docs/docs/pyproject.md
Comment (DustinMoriarty) [inline docs/docs/pyproject.md:256]: It looks like this is just the automated pre-commit hook for white space.

**Comment by DustinMoriarty** [comment]
Well I did do some additional changes to improve the CI help documentation. The FreeBSD check passed this time. Given the nature of that issue intermittently not being able to find a file I wonder if the Cirrus FreeBSD service is running low on storage. I guess all is well for this BR, but maybe something to look into.

**Comment by sinoroc** [comment]
I have seen multiple reports of CI failing on FreeBSD. I would not worry about this too much. Usually just re-triggering the build solves it.

**Comment by DustinMoriarty** [comment]
Sounds good.  @abn or @steveYeah : could you please review? Is this ready to merge?

**Review by neersighted — CHANGES_REQUESTED** [review]
This has some conflicts -- LGTM once they're solved.

**Comment by DustinMoriarty** [comment]
> This has some conflicts -- LGTM once they're solved.

@neersighted : I merged from the upstream master and had no merge conflicts. It looks like it passes the pre-commit checks. Can you please approve the workflow pipeline to run? Thank you.

File: docs/managing-environments.md
Comment (neersighted) [inline docs/managing-environments.md:134]:
```suggestion
poetry env remove python3.6 python3.7 python3.8
```

This looks to be by executable and not by name.

**Comment by DustinMoriarty** [comment]
@neersighted : It looks like this one passed all pipeline checks. I believe that all suggested changes are addressed. Can you please approve the merge? Thank you.

File: poetry/console/commands/env/remove.py
Comment (neersighted) [inline poetry/console/commands/env/remove.py:38]:
```suggestion
        # TODO: refactor env.py to allow removal with one loop
        for python in pythons:
```

Let's drop a `TODO` here to clean this up a little better.

**Review by neersighted — CHANGES_REQUESTED** [review]
I'd like a couple changes to the command descriptions, and I'm not 100% thrilled with the less-than-DRY approach. However, the code is simple to understand. Slight refactoring in `env.py` is needed to DRY this up -- that can be left as a future task.

File: docs/managing-environments.md
Comment (DustinMoriarty) [reply docs/managing-environments.md:134]: Sounds good if that is the best practice. I always do it the other way.  Good to know!

**Comment by DustinMoriarty** [comment]
@neersighted : All suggestions should be resolved now. Thank you for the suggestions.

**Comment by neersighted** [comment]
Awesome! I'm going to squash your changes, but keep in mind Github lets you batch suggested edits into one commit. I'd have to squash here anyway, but it's worth keeping in mind if you do like using the web interface.

**Review by neersighted — APPROVED** [review]

**Pull request merged**
